### ISSUE TITLE
Introduce periodic process stats (mem, page faults) into systrace

### ISSUE THREAD

**Issue by primiano** [catapult-project/catapult#3437]
Tracker bug to introduce the notion of periodic process stats into the systrace.
In the very essence, I'd like to get top-level counters (memory, page faults, oom scores) samples and exposed in the trace, similary to what we do in chrome for memory, e.g.:

![image](https://cloud.githubusercontent.com/assets/7137473/24245550/3b65d59e-0fbb-11e7-979a-28e26fe07ff4.png)

The plan is to:

- Create a helper binary that periodically samples these stats into a JSON file
- Have an optional agent that can be started together with the atrace agent by systrace which pushes and run the aforementioned binary
- Have an importer which reads the output of that binary and merges back into the trace model

This binary can be also used to keep track of process names as they come up and die during the trace (Right now we only sample them at the end)

**Comment by primiano** [comment]
+@ChrisCraik

benm@ is on the importer side.

Draft CL for the helper daemon here:
https://codereview.chromium.org/2724443004/

**Comment by ChrisCraik** [comment]
Nice, the general approach looks great! Somewhat similar to how Android Studio captures/displays similar information in their UI.

How expensive is writing out the JSON? Would it be less intrusive to stash it in memory until it's dump time, or is it too large?

**Comment by primiano** [comment]
Yeah I was considering that. The expensive part is really the computatio.
The kernel<>userspace interface is idiotic and the only way  to get memory details involves parsing a *human readable* file, blah. It takes ~1.5-2s to fully dump (i.e. with the -m option) memory stats for all processes in the system. Most of that is blocked on waiting on I/O on the pseudo /proc filesystem. Writing itself is not that much (say ~50k per snapshot more or less).

I could stash in memory and dump at the end, but I worry that if I do that then I can interfere too much with the low memory killer (as my helper process footprint would grow without bounds).
Instead, by writing a file incrementally, my expectation is that the write overhead is minimal (there is no fdatasync nor I/O barrier involved, so the write() will just hit the pagecache) but at the same time, if we write a lot of stuff, the kernel will actually commit the pagecache back to the file, and we'd not run into the long-trace problem.

**Comment by ChrisCraik** [comment]
+@jreck

**Comment by jreck** [comment]
My initial thoughts are a combination of this is going to heavily skew trace results as the polling check is going to be a heavy, bursty workload. If we do want this it's probably better as a daemon inside atrace so that atrace remains the single entry-point into tracing. But I'm going to first follow up with the systems team to see if we could just add some of this to ftrace instead so it's event-driven instead of polling.

**Comment by jreck** [comment]
page faults are already in the ftrace buffer if you have sched enabled I believe and trace-viewer will highlight it (it's the bright yellow color in the sched bar). Is there anything else necessary there?

**Comment by primiano** [comment]
> My initial thoughts are a combination of this is going to heavily skew trace results as the polling check is going to be a heavy, bursty workload.
True, we have the same issue in chrome. The idea is that is an extra flag that you have to pass explicitly (read: not enabled by default) to get these periodic snapshots.

> If we do want this it's probably better as a daemon inside atrace so that atrace remains the single entry-point into tracing
What is the benefit? Downsides I see: this increases coupling, mainteinance cost of atrace, and reduces stability of atrace (if something goes wrong in the stats collecting process, the entire trace will go bankrupt). 

> But I'm going to first follow up with the systems team to see if we could just add some of this to ftrace instead so it's event-driven instead of polling.
Unfortunately, I don't that even-driven applies memory (purely from a conceptually viewpoint). Virtually any instruction changes the memory footprint of a process, and we definitely don't want to spam the trace with memory updates every other instruction. Also, I looked into the kernel implementation that produces those stats, and some of those stats (in particular PSS) require extensive computation. So I can't see how they could be event-driven.

> page faults are already in the ftrace buffer if you have sched enabled I believe and trace-viewer will highlight it (it's the bright yellow color in the sched bar). Is there anything else necessary there?

What I want to achieve here is: given a range in the timeline, I want to see how many soft/hard page faults each process was subjected to. The snapshot ui linked above seemed the right way to get that. (Anyhow, reading pagefaults from /proc/pid/statm is a zero cost operation given that I have to read that file for memory purposes. The real heavy stuff comes from parsing /proc/pid/smaps, and I am afraid there is no way around that if we want memory dumps.)

**Comment by jreck** [comment]
systrace.py is but 1 of 4 mechanisms that I'm currently aware of for initiating a trace (and 1 of 5 for dumping a trace), so we would need to re-implement this in a few other places if we want this data in the other pipelines. That's why I was suggesting it be done as an atrace option instead since atrace is the common entry point that everything shares.

**Comment by ChrisCraik** [comment]
It might make sense to build it into both. Building into atrace enables other means of capture (such as host-less), and simplifies the job of the tracer long term. Building into the python tracer allows us to use the feature immediately.

If the output format isn't easy to agree upon (here, seems like something that might be extensible and grow in complexity over time). For something like dumping process names for thread name fixups, that's a clearer candidate for atrace.

**Comment by gkraynov** [comment]
Implemented optional memory maps with periodic memory dumps. Example below has produced by
`systrace --process-dump --process-dump-full=system_server --process-dump-mmaps`

![systracemmaps](https://user-images.githubusercontent.com/7749127/28613900-c7702d40-71eb-11e7-9723-8d19ef9dd0f4.png)
